From b70fc022cae3efec16efb2b46a0440d3e6dd3bdc Mon Sep 17 00:00:00 2001
From: Matt Breedlove <breedlove.matt@gmail.com>
Date: Wed, 5 Aug 2015 23:36:22 +0100
Subject: [PATCH 12/22] Work around {un-,too-late-}resolved __EH_FRAME_BEGIN__

Here's the dwarf2 patch I'm using.  With the change to checking
__LIBGCC_EH_FRAME_SECTION_NAME__, an undefined reference to
__EH_FRAME_BEGIN__ gets emitted within __main.o.  The actual symbol
for it is defined within cygming-crtbegin.c (crtbegin.o) which is
listed before __main.o when linking libgcc leaving the undefined
reference unresolved.  This restores behavior to the 4.9 branch only
for this specific instance (confirmed by comparing pre-processed
output).  Will update with bugtraq info.

Relevant references:

https://gcc.gnu.org/ml/gcc-patches/2014-09/msg00409.html
http://gcc.gcc.gnu.narkive.com/RMY4RcbE/asm-output-section-name-cleanup-patch-problem

Posted to mingw-w64 ML:

https://sourceforge.net/p/mingw-w64/mailman/message/34101954/
https://gcc.gnu.org/ml/gcc-patches/2015-05/msg01212.html
---
 libgcc/libgcc2.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/libgcc/libgcc2.c b/libgcc/libgcc2.c
index c737620..e5ab209 100644
--- a/libgcc/libgcc2.c
+++ b/libgcc/libgcc2.c
@@ -2211,7 +2211,7 @@ TRANSFER_FROM_TRAMPOLINE
 /* Some ELF crosses use crtstuff.c to provide __CTOR_LIST__, but use this
    code to run constructors.  In that case, we need to handle EH here, too.  */
 
-#ifdef __LIBGCC_EH_FRAME_SECTION_NAME__
+#ifdef EH_FRAME_SECTION_NAME
 #include "unwind-dw2-fde.h"
 extern unsigned char __EH_FRAME_BEGIN__[];
 #endif
@@ -2231,7 +2231,7 @@ __do_global_dtors (void)
       (*(p-1)) ();
     }
 #endif
-#if defined (__LIBGCC_EH_FRAME_SECTION_NAME__) && !defined (HAS_INIT_SECTION)
+#if defined (EH_FRAME_SECTION_NAME) && !defined (HAS_INIT_SECTION)
   {
     static int completed = 0;
     if (! completed)
@@ -2250,7 +2250,7 @@ __do_global_dtors (void)
 void
 __do_global_ctors (void)
 {
-#ifdef __LIBGCC_EH_FRAME_SECTION_NAME__
+#ifdef EH_FRAME_SECTION_NAME
   {
     static struct object object;
     __register_frame_info (__EH_FRAME_BEGIN__, &object);
-- 
2.5.0

